### PR TITLE
Update 2.1.0 again

### PR DESCRIPTION
**Pull request by hmaarrfk**
@jakirkham the same error happened again.

The CI should show it too:

```
  -- Found Threads: TRUE
  CMake Error at cmake/public/cuda.cmake:65 (message):
    Found two conflicting CUDA installs:

    V12.0.76 in
    '/home/conda/feedstock_root/build_artifacts/pytorch-recipe_1699320420794/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeh/targets/x86_64-linux/include'
    and

    V12.0.76 in
    '/home/conda/feedstock_root/build_artifacts/pytorch-recipe_1699320420794/_build_env/targets/x86_64-linux/include'
  Call Stack (most recent call first):
    cmake/Dependencies.cmake:44 (include)
    CMakeLists.txt:722 (include)
```
Checklist
* [ ] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [ ] Bumped the build number (if the version is unchanged)
* [ ] Reset the build number to `0` (if the version changed)
* [ ] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [ ] Ensured the license file is being packaged.

### CODE REVIEW

**Comment by jakirkham** [comment]
Thanks Mark! 🙏

It appears this is due to a change introduced in [PyTorch 2.1.0's own CTK detection logic]( https://github.com/pytorch/pytorch/pull/82695/files#diff-8e5cb190cc46be808993381a31fe9c027705d356b6bc0460368c0310ae82b273R61-R66 ). Unfortunately this check is problematic

In the Conda case, we implement a splayed layout. This means build tools (like those listed in `requirements/build`) live in one path and libraries that are linked to (like those listed in `requirements/host`) live in another path. This is common when supporting cross-compilation as we do with Conda. However this means there are cases where we do need to use some things from both paths for different reasons

It looks like @ax3l ran into this issue in the HPC SDK use case and proposed a fix ( https://github.com/pytorch/pytorch/pull/108932 ). Idk if that will work for us

In terms of our needs here, maybe just patching the check out altogether would be reasonable to get the build working

Perhaps this would be a good opportunity to discuss with @peterbell10 whether we can come up with a better check in PyTorch that works for splayed layout use cases

**Comment by hmaarrfk** [comment]
ok lets try again, i took a brute force approach because we can do a bit of meta building

File: recipe/meta.yaml [hunk 1]
```diff
@@ -25,6 +25,11 @@ source:
   git_tag: v{{ version }}
   patches:
     - nvml.patch  # [cuda_compiler_version == "11.2"]
+    # The patch below is probably OK for other versions too, but it is untested with them
+    # https://github.com/conda-forge/pytorch-cpu-feedstock/pull/203#issuecomment-1797352452
+    # https://github.com/pytorch/pytorch/pull/82695/files#diff-8e5cb190cc46be808993381a31fe9c027705d356b6bc0460368c0310ae82b273R61-R66
+    # https://github.com/pytorch/pytorch/pull/108932
+    - 0001-Allow-splayed-layouts.patch  # [cuda_compiler_version == "12.0"]
```
Comment (jakirkham) [inline recipe/meta.yaml:32]: Think we might need this

```suggestion
    - patches/0001-Allow-splayed-layouts.patch  # [cuda_compiler_version == "12.0"]
```

Comment (hmaarrfk) [reply recipe/meta.yaml:32]: thanks . i got it locally. i cleaned up a little more.

**Comment by hmaarrfk** [comment]
It now fails with:
```
  CMake Error at cmake/public/cuda.cmake:64 (message):
    Failed to find nvToolsExt
  Call Stack (most recent call first):
    cmake/Dependencies.cmake:44 (include)
    CMakeLists.txt:722 (include)


```

File: recipe/meta.yaml (hunk 1 above)
Comment (jakirkham) [reply recipe/meta.yaml:32]: Thanks Mark! 🙏

**Comment by jakirkham** [comment]
Thanks Mark! 🙏

Think that refers to [this check
added in the same PyTorch PR]( https://github.com/pytorch/pytorch/pull/82695/files#diff-8e5cb190cc46be808993381a31fe9c027705d356b6bc0460368c0310ae82b273R68-R70 )

As noted under [the `CUDA::nvToolsExt` doc in CMake]( https://cmake.org/cmake/help/latest/module/FindCUDAToolkit.html#cuda-toolkit-nvtoolsext ), this is a deprecated target by CMake (and NVIDIA) as it comes from NVTX 2, NVTX 3 has superseded it

Am guessing PyTorch doesn't use NVTX 2 (as this check was new in that PR). Meaning that this was purely a build configuration check

So think we can remove those lines as well

**Comment by hmaarrfk** [comment]
what about this:
https://github.com/pytorch/pytorch/pull/82695/files#diff-8e5cb190cc46be808993381a31fe9c027705d356b6bc0460368c0310ae82b273R218

**Comment by hmaarrfk** [comment]
can i remove it too?

File: recipe/build_pytorch.sh [hunk 2]
```diff
@@ -128,20 +128,28 @@ if [[ ${cuda_compiler_version} != "None" ]]; then
...
+        export CUDA_TOOLKIT_ROOT_DIR=$CUDA_HOME
     elif [[ ${cuda_compiler_version} == 11.2 ]]; then
         export TORCH_CUDA_ARCH_LIST="3.5;5.0;6.0;6.1;7.0;7.5;8.0;8.6+PTX"
+        export CUDA_TOOLKIT_ROOT_DIR=$CUDA_HOME
     elif [[ ${cuda_compiler_version} == 11.8 ]]; then
         export TORCH_CUDA_ARCH_LIST="3.5;5.0;6.0;6.1;7.0;7.5;8.0;8.6;8.9+PTX"
+        export CUDA_TOOLKIT_ROOT_DIR=$CUDA_HOME
     elif [[ ${cuda_compiler_version} == 12.0 ]]; then
         export TORCH_CUDA_ARCH_LIST="5.0;6.0;6.1;7.0;7.5;8.0;8.6;8.9;9.0+PTX"
+        # CUDA_TOOLKIT_ROOT_DIR no longer needed for 12.0
```
Comment (jakirkham) [inline recipe/build_pytorch.sh:152]:
```suggestion
        # $CUDA_HOME not set in CUDA 12.0. Using $PREFIX
        export CUDA_TOOLKIT_ROOT_DIR="${PREFIX}"
```

**Review by jakirkham — COMMENTED** [review]
Hmm...think you make a good point

In looking through the code see [some references to NVTX 2]( https://github.com/pytorch/pytorch/blob/ddfe5725342b0c0f707222879ca9dac305f97210/torch/csrc/profiler/stubs/cuda.cpp#L3 )

We could override [where PyTorch looks for NVTX]( https://github.com/pytorch/pytorch/blob/ddfe5725342b0c0f707222879ca9dac305f97210/cmake/Modules/FindCUDAToolkit.cmake#L1051 ), but maybe it is best to point PyTorch to where the bulk of the CTK content is in the native compilation case

It's possible this will need tweaking further in the build, but hopefully this unblocks configuration

**Comment by hmaarrfk** [comment]
ok well i pushed my changes, feel free to push anything if you can. going to slee....

**Comment by jakirkham** [comment]
Thanks Mark! 🙏

Have a good night

**Comment by jakirkham** [comment]
It looks like Peter added a fix upstream ( https://github.com/pytorch/pytorch/pull/113174 ). Thanks Peter! 🙏

Maybe we can give that a try

**Comment by hmaarrfk** [comment]
it doesn't address the nvtools issue

**Comment by Tobias-Fischer** [comment]
There are two upsteam PRs that solve the nvtools issue; I am not sure which one is preferable. See https://github.com/pytorch/pytorch/issues/101135 and PRs https://github.com/pytorch/pytorch/pull/97582 and https://github.com/pytorch/pytorch/pull/106763

File: recipe/build_pytorch.sh [hunk 3]
```diff
@@ -128,20 +128,29 @@ if [[ ${cuda_compiler_version} != "None" ]]; then
...
     elif [[ ${cuda_compiler_version} == 11.2 ]]; then
         export TORCH_CUDA_ARCH_LIST="3.5;5.0;6.0;6.1;7.0;7.5;8.0;8.6+PTX"
+        export CUDA_TOOLKIT_ROOT_DIR=$CUDA_HOME
     elif [[ ${cuda_compiler_version} == 11.8 ]]; then
         export TORCH_CUDA_ARCH_LIST="3.5;5.0;6.0;6.1;7.0;7.5;8.0;8.6;8.9+PTX"
+        export CUDA_TOOLKIT_ROOT_DIR=$CUDA_HOME
     elif [[ ${cuda_compiler_version} == 12.0 ]]; then
         export TORCH_CUDA_ARCH_LIST="5.0;6.0;6.1;7.0;7.5;8.0;8.6;8.9;9.0+PTX"
+        # $CUDA_HOME not set in CUDA 12.0. Using $PREFIX
+        export CUDA_TOOLKIT_ROOT_DIR="${PREFIX}"
```
Comment (jakirkham) [inline recipe/build_pytorch.sh:153]: Maybe we need to point PyTorch to the particular target directory. Usually this hasn't been needed with CMake builds, but there is at least one case where we had to go this route ( https://github.com/conda-forge/kaldi-feedstock/pull/39 )

Here's [the layout logic]( https://github.com/conda-forge/cuda-nvcc-feedstock/blob/394f3817decd9316269b5a40e5076b8bcc94dc33/recipe/activate.sh#L5-L7 ) for context

```suggestion
        if [[ "${target_platform}" == "linux-64" ]]; then
            export CUDA_TOOLKIT_ROOT_DIR="${PREFIX}/targets/x86_64-linux"
        elif [[ "${target_platform}" == "linux-ppc64le" ]]; then
            export CUDA_TOOLKIT_ROOT_DIR="${PREFIX}/targets/ppc64le-linux"
        elif [[ "${target_platform}" == "linux-aarch64" ]]; then
            export CUDA_TOOLKIT_ROOT_DIR="${PREFIX}/targets/sbsa-linux"
        else
            echo "Unsupported architecture by CUDA"
            exit 1
        fi
```

Comment (jakirkham) [reply recipe/build_pytorch.sh:153]: Also if we would rather just do the `linux-64` logic, that's fine. Just wanted to spell out the other architectures for completeness

**Comment by hmaarrfk** [comment]
ok builds are incoming:

My test was to use 
```python
import torch
a = torch.randn(1024 * 1024 * 1024, device='cuda')
a + 1
```

and watch the memory on my cuda device grow using nvtop

File: recipe/meta.yaml [hunk 4]
```diff
@@ -33,6 +44,7 @@ build:
   skip: true  # [cuda_compiler_version == "10.2"]
   skip: true  # [cuda_compiler_version == "11.0"]
   skip: true  # [cuda_compiler_version == "11.1"]
+  skip: true  # [cuda_compiler_version != "12.0"]
```
Comment (jakirkham) [inline recipe/meta.yaml:47]: Guessing this is a temporary measure to reign in builds offline. Is that right?

**Comment by jakirkham** [comment]
Nice work Mark! 🥳

@peterbell10 do the patches here look upstreamable to you? Or are there similar approaches that upstream could take that would alleviate the need for these?

File: recipe/meta.yaml (hunk 4 above)
Comment (hmaarrfk) [reply recipe/meta.yaml:47]: yes. I can add a note about it. but I'm not thrilled about using 8OSX CPUs for 8-12 hours for this.

Comment (Tobias-Fischer) [reply recipe/meta.yaml:47]: Do you mean the CI CPUs? If so, can't we stop the Azure builds manually? Or do you need help with local builds?

Comment (hmaarrfk) [reply recipe/meta.yaml:47]: I don't know how to stop the build manually. I just feel that somebody will remember to enable the osx builds.....

Comment (Tobias-Fischer) [reply recipe/meta.yaml:47]: Agreed :)

**Review by Tobias-Fischer — APPROVED** [review]

**Comment by hmaarrfk** [comment]
[log_files.zip](https://github.com/conda-forge/pytorch-cpu-feedstock/files/13315316/log_files.zip)

**Comment by jakirkham** [comment]
Were Linux ARM packages built as well?

**Pull request merged**